File: Merge Two Sorted Lists/Merge two sorted list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *first = list1, *second = list2;
        ListNode *head = NULL, *tail = NULL;
        
        while (first != NULL || second != NULL) {
            if (first != NULL && second != NULL) {
                if (first->val < second->val) {
                    tail = addAtTheEnd(tail, first->val);
                    first = first->next;
                }
                else {
                    tail = addAtTheEnd(tail, second->val);
                    second = second->next;
                }
            }
            else if (first != NULL) {
                tail = addAtTheEnd(tail, first->val);  
                first = first->next;
            }
            else {
                tail = addAtTheEnd(tail, second->val);
                second = second->next;
            }
            
            if (head == NULL) {
                head = tail;
            }
        }
        
        return head;
    }
    
    ListNode* addAtTheEnd(ListNode *tail, int data) {
        ListNode *ntbi = new ListNode;
        ntbi->val = data;
        ntbi->next = NULL;
        
        if (tail != NULL) {
            tail->next = ntbi;
        }
        return ntbi;
    }
};

File: Merge Two Sorted Lists/merge two sorted list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        p1 = list1
        p2 = list2
        tail = None
        head = None
        
        while (p1 is not None or p2 is not None):
            data = None
            if (p1 is not None and p2 is not None):
                if (p1.val <= p2.val):
                    data = p1.val
                    p1 = p1.next
                else:
                    data = p2.val
                    p2 = p2.next
            elif (p1 is not None):
                data = p1.val
                p1 = p1.next
            else:
                data = p2.val
                p2 = p2.next
            
            if (tail is None):
                head = self.insertAtEnd(tail, data)
                tail = head
            else:
                tail = self.insertAtEnd(tail, data)
        
        return head
                
        
    
    def insertAtEnd(self, tail, data):
        ntbi = ListNode(data)
        if (tail  is not None):
            tail.next = ntbi
        return ntbi
        